### PR TITLE
Support DotNet auto-instrumentation

### PR DESCRIPTION
**Pull request by avadhut123pisal**
This PR adds a support for auto-instrumentation of the  .Net applications. 
Currently, dot net auto-instrumentation image is using the release [opentelemetry-dotnet-instrumentation v0.2.0-beta.1](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/releases/tag/v0.2.0-beta.1).

Following environment variables are injected to the application container to enable the auto-instrumentation.

```
DOTNET_ADDITIONAL_DEPS=%InstallationLocation%/AdditionalDeps
DOTNET_SHARED_STORE=%InstallationLocation%/store
DOTNET_STARTUP_HOOKS=%InstallationLocation%/netcoreapp3.1/OpenTelemetry.AutoInstrumentation.StartupHook.dll
```

Resolves [#756](https://github.com/open-telemetry/opentelemetry-operator/issues/756)

### CODE REVIEW

File: autoinstrumentation/dotnet/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# To build one auto-instrumentation image for dotnet, please:
+#  - Download your dotnet auto-instrumentation artefacts to `/autoinstrumentation` directory. This is required as when instrumenting the pod,
+#    one init container will be created to copy the files to your app's container.
```
[the remark this answers is not in the document]
Comment (avadhut123pisal) [reply autoinstrumentation/dotnet/Dockerfile:3]: Following environment variables are injected to the application container to enable the auto-instrumentation.
```
DOTNET_ADDITIONAL_DEPS=%InstallationLocation%/AdditionalDeps
DOTNET_SHARED_STORE=%InstallationLocation%/store
DOTNET_STARTUP_HOOKS=%InstallationLocation%/netcoreapp3.1/OpenTelemetry.AutoInstrumentation.StartupHook.dll
```

[the remark this answers is not in the document]
Comment (avadhut123pisal) [reply autoinstrumentation/dotnet/Dockerfile:3]: Added this to the dockerfile

[the remark this answers is not in the document]
Comment (avadhut123pisal) [reply autoinstrumentation/dotnet/Dockerfile:3]: > Following environment variables are injected to the application container to enable the auto-instrumentation.
> 
> ```
> DOTNET_ADDITIONAL_DEPS=%InstallationLocation%/AdditionalDeps
> DOTNET_SHARED_STORE=%InstallationLocation%/store
> DOTNET_STARTUP_HOOKS=%InstallationLocation%/netcoreapp3.1/OpenTelemetry.AutoInstrumentation.StartupHook.dll
> ```

Added this to the dockerfile.

**Comment by avadhut123pisal** [comment]
e2e test case is failing for newly added dot-net auto-instrumentation with the error in pulling the docker image `ghcr.io/open-telemetry/opentelemetry-operator/autoinstrumentation-dotnet:0.0.1-alpha`. But, until the review gets completed and it is merged, this image will not be available. So, how to deal with this ?

**Comment by pavolloffay** [comment]
> But, until the review gets completed and it is merged, this image will not be available. So, how to deal with this ?

Please open a separate PR just to build and publish the image

**Comment by avadhut123pisal** [comment]
> > But, until the review gets completed and it is merged, this image will not be available. So, how to deal with this ?
> 
> Please open a separate PR just to build and publish the image

I have opened a separate PR to build and publish the image [#989](https://github.com/open-telemetry/opentelemetry-operator/pull/989)

**Comment by avadhut123pisal** [comment]
e2e test case for dotnet auto-instrumentation is failing because it is facing an issue in pulling the image. I think, the package got published with private visibility.

**Comment by pavolloffay** [comment]
what is the binary file `.DS_Store` in this PR?

File: tests/e2e/instrumentation-dotnet-multicontainer/00-install-instrumentation.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    endpoint: http://localhost:4317
+  propagators:
+    - jaeger
+    - b3
+  sampler:
+    type: parentbased_traceidratio
+    argument: "0.25"
+  dotnet:
+    env:
+    - name: OTEL_DOTNET_AUTO_TRACES_ENABLED_INSTRUMENTATIONS
```
Comment (pavolloffay) [inline tests/e2e/instrumentation-dotnet-multicontainer/00-install-instrumentation.yaml:29]: is this required? What happens if the variable is empty?

Comment (avadhut123pisal) [reply tests/e2e/instrumentation-dotnet-multicontainer/00-install-instrumentation.yaml:29]: The PR [#932 Enable all trace instrumentations by default (simplified)](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/932) for enabling the all trace instrumentation by default was merged few days back. 

So, `OTEL_DOTNET_AUTO_TRACES_ENABLED_INSTRUMENTATIONS` is required to be specified for the release [opentelemetry-dotnet-instrumentation v0.2.0-beta.1](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/releases/tag/v0.2.0-beta.1).

If it is empty, then applications runs but trace instrumentation does not work.

**Comment by avadhut123pisal** [comment]
> 

My bad. This file is not required. I will delete it and commit.

**Comment by avadhut123pisal** [comment]
> what is the binary file `.DS_Store` in this PR?

My bad. This file is not required. I will delete it and commit.

File: tests/e2e/instrumentation-dotnet-multicontainer/00-install-instrumentation.yaml (hunk 2 above)
Comment (pavolloffay) [reply tests/e2e/instrumentation-dotnet-multicontainer/00-install-instrumentation.yaml:29]: Will this change in the future? If no the operator should set this variable on the behalf of the user.

Comment (avadhut123pisal) [reply tests/e2e/instrumentation-dotnet-multicontainer/00-install-instrumentation.yaml:29]: Previously, this env variable's name was `OTEL_DOTNET_TRACER_INSTRUMENTATIONS `. In v0.2.0-beta.1, it is renamed to `OTEL_DOTNET_AUTO_TRACES_ENABLED_INSTRUMENTATIONS `. 

As of now, we can set this env value to `AspNet,HttpClient,SqlClient` but in the next release of the dotnet auto-instrumentation `Npgsql, MySqlData` will be added.

Comment (pavolloffay) [reply tests/e2e/instrumentation-dotnet-multicontainer/00-install-instrumentation.yaml:29]: We should definitely document this in the readme and the operator could set the variable (if it is not set) in the defaulting webhook

Comment (avadhut123pisal) [reply tests/e2e/instrumentation-dotnet-multicontainer/00-install-instrumentation.yaml]: Done. I have pushed the changes.

**Review by pavolloffay — APPROVED** [review]

**Pull request merged**